framework/src/auxkernels/MaterialStdVectorAux.C

```diff
@@ -15,27 +15,24 @@
 #include "MaterialStdVectorAux.h"
 
 template<>
-InputParameters validParams<MaterialStdVectorAux>()
+InputParameters validParams< MaterialStdVectorAux>()
 {
-  InputParameters params = validParams<MaterialAuxBase<std::vector<Real> > >();
-  params.addParam<unsigned int>("index", 0, "The index to consider for this kernel");
-  params.addClassDescription("Extracts a component of a material's std::vector<Real> to an aux variable.  If the std::vector is not of sufficient size then zero is returned");
-
+  InputParameters params = validParams<MaterialStdVectorAuxBase<> >();
+  params.addClassDescription("Extracts a component of a material type std::vector<Real> to an aux variable.  If the std::vector is not of sufficient size then zero is returned");
   return params;
 }
 
-MaterialStdVectorAux::MaterialStdVectorAux(const std::string & name, InputParameters parameters) :
-  MaterialAuxBase<std::vector<Real> >(name, parameters),
-    _index(getParam<unsigned int>("index"))
+ MaterialStdVectorAux:: MaterialStdVectorAux(const std::string & name, InputParameters parameters) :
+  MaterialStdVectorAuxBase<Real>(name, parameters)
 {
 }
 
-MaterialStdVectorAux::~MaterialStdVectorAux()
+ MaterialStdVectorAux::~ MaterialStdVectorAux()
 {
 }
 
 Real
-MaterialStdVectorAux::computeValue()
+ MaterialStdVectorAux::computeValue()
 {
   mooseAssert(_prop[_qp].size() > _index, "MaterialStdVectorAux: You chose to extract component " << _index << " but your Material property only has size " << _prop[_qp].size());
   return _factor * _prop[_qp][_index] + _offset;
```

Comment:
That is not what I was thinking, but that is better. I was just thinking of
calling a method inside of computeValue, but it is probably better design
to create a new virtual that is defined in the child classes.

We need the C++11 `final` for computeValue

On Thu, Jul 2, 2015 at 8:43 AM, Daniel Schwen <notifications@github.com>
wrote:

> In framework/src/auxkernels/MaterialStdVectorAux.C
> <https://github.com/idaholab/moose/pull/5331#discussion_r33784621>:
>
> >  {
> >  }
> >
> >  Real
> > -MaterialStdVectorAux::computeValue()
> > + MaterialStdVectorAux::computeValue()
> >  {
> >    mooseAssert(_prop[_qp].size() > _index, "MaterialStdVectorAux: You chose to extract component " << _index << " but your Material property only has size " << _prop[_qp].size());
> >    return _factor * _prop[_qp][_index] + _offset;
>
> Oh, wait, the base class method still needs to be computeValue! The
> derived classes need to implement a new methos. I'd have called it
> computeComponent.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/5331/files#r33784621>.
>